### ISSUE TITLE
East-west gateway without external IP

### ISSUE THREAD

**Issue by fizos** [istio/istio#49045]
I am using Istio 1.17.3 and I want to deploy a multicluster mesh where each cluster is in a different network. I am following the [Install Multi-Primary on different networks 29](https://istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/) guide and my environment doesn’t support external load balancers. So when I install an east-west gateway it doesn’t get an external IP and I cannot get any cross-cluster traffic.

Is there any workaround for this (e.g. accessing the gateway using a nodePort)?

**Comment by stevenctl** [comment]
Yes, we do support NodePort for multi-network. 

The docs have you use the gen-eastwest-gateway.sh. After generating the YAML using that command, you'd need to edit one section

```
  values:
    gateways:
      istio-ingressgateway:
        injectionTemplate: gateway
        type: NodePort # add this
```

**Comment by fizos** [comment]
very appreciate for you help 😄 

I currently need to form a large service mesh cluster using a multi-primary, multi-network mode with three clusters. Is it necessary for each cluster to configure the access credentials of the other two clusters' API-Server? Does Istio support this configuration ? Do East-West gateways also need to be configured for mutual communication between each pair? 
Is there a best practice in the community for three clusters with multi-primary and multi-network?

**Comment by stevenctl** [comment]
Istio's multi-cluster design currently requires API server access to be shared with each primary cluster. 

[Primary-Remote](https://istio.io/latest/docs/setup/install/multicluster/primary-remote_multi-network/) would only require supplying the remote clusters' credentials to each primary (you could have 1 primary, 2 remote). 

Another option would be more manual: creating `WorkloadEntry` resources and `Gateway` resources by hand that represent resources that exist in other clusters. An example can be found in this [integration test](https://github.com/istio/istio/blob/master/tests/integration/pilot/workloadentry_test.go). A `Gateway` represents a remote eastwest gateway, we say the address `1.2.3.4:15443` can be used to reach any endpoint on `network-foo`.  Then we create a `WorkloadEntry` on `network-foo` with no address. A service selecting that `WorkloadEntry` would go through the gateway. 

The caveats being:

1. You don't get automated updates if you have no workloads behind that "virtual" workloadentry.
2. You don't get automated updates for the Gateway resource
3. The WorkloadEntry may not have the correct weight for the # of workloads in the remote network. 

TL;DR we have APIs that can be used to configure some cross-network traffic without sharing API server credentials, but you lose most of the automation that Istio provides.

**Comment by fizos** [comment]
> Yes, we do support NodePort for multi-network.
> 
> The docs have you use the gen-eastwest-gateway.sh. After generating the YAML using that command, you'd need to edit one section
> 
> ```
>   values:
>     gateways:
>       istio-ingressgateway:
>         injectionTemplate: gateway
>         type: NodePort # add this
> ```
> 
> The multi-cluster service discovery machinery is setup to handle node port.

I have added the type NodePort in eastwestgateway istiooperator，and exposed the gateway with Gateway CRD.
But it seems that the traffic from the west cluster did not find the eastwestgateway in east cluster by nodeport
```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: eastwest
spec:
  revision: ""
  profile: empty
  tag: 1.17.3
  components:
    ingressGateways:
    - name: istio-eastwestgateway
      label:
        istio: eastwestgateway
        app: istio-eastwestgateway
        topology.istio.io/network: east-network
      enabled: true
      k8s:
        env:
        # traffic through this gateway should be routed inside the network
        - name: ISTIO_META_REQUESTED_NETWORK_VIEW
          value: east-network
        service:
          ports:
          - name: status-port
            port: 15021
            targetPort: 15021
          - name: tls
            port: 15443
            targetPort: 15443
          - name: tls-istiod
            port: 15012
            targetPort: 15012
          - name: tls-webhook
            port: 15017
            targetPort: 15017
  values:
    gateways:
      istio-ingressgateway:
        injectionTemplate: gateway
        type: NodePort
    global:
      network: east-network
```
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: cross-network-gateway
  namespace: istio-system
spec:
  selector:
    istio: eastwestgateway
  servers:
  - port:
      number: 15443
      name: tls
      protocol: TLS
    tls:
      mode: AUTO_PASSTHROUGH
    hosts:
    - "*.local"
```
Besides the eastwestgateway generated by gen-eastwest-gateway.sh did not contain mesh-id info, Is this right?

**Comment by fizos** [comment]
When I change the type of the eastwest gateway to nodeport and specify the external IP as the node IP, everything works fine. Now, I want to use nginx to provide a layer 4 proxy to proxy the eastwest gateway nodeport for high availability. However, when I enable this configuration, the eastwest gateway starts to behave abnormally.
![image](https://github.com/istio/istio/assets/96819554/5d038995-d591-419c-9a27-d39cd1807f6b)
externalIP：nginx server ip
![image](https://github.com/istio/istio/assets/96819554/8da373d6-aa13-45be-ae76-769b3c8d0d8f)
proxy_pass nodeip

Is something wrong？

**Comment by lgy1027** [comment]
Have you solved the problem? I don’t have an external IP either.

**Comment by fizos** [comment]
Yes , I have solved this problem, you can install a loadbalancer like nginx outside the cluster and forward the 15443 port

**Comment by lgy1027** [comment]
> Yes , I have solved this problem, you can install a loadbalancer like nginx outside the cluster and forward the 15443 port

I am deploying an istio cluster using a non-flat multi-master solution. If nideport is used to expose the east-west gateway, does this address need to be set to istio of the opposite cluster? Otherwise, the applications deployed by istio cannot communicate with each other.

**Comment by fizos** [comment]
No，with the remote-cluster-secret， the istio can discover the service in  the opposite cluster.
What you have to do is to deploy the secets in the opposite cluster.

**Comment by marshell0** [comment]
I tested the type NodePort in eastwestgateway istiooperator, and couldn't get the request route the the opposite, and I checked the proxy-config endpoint, It did generated a item point to an internal ip:5000.
Then I tried metallab, eastwestgateway was assigned an external IP, and proxy-config endpoint shows that an exteral IP and port 15443 item generated, but the request still couln't be routed to the other side, but the assinged external IP of ingressgateway works well.
If I set this external IP to an IP of cluster nodes, everything goes well, the verification application helloworld could get reponses with V1 and V2 randomly
BTW, I have a custom trustDomain set my istio install, I have set trustDomainAlias for this situaion.
A little bit confused

**Comment by vnhirenko** [comment]
@fizos , if we use node port for istio east-west gateway an we have an external load balancer to route traffic to published on node ips, how istio on another cluster will know the hostname or ip of the  external load balancer? Would it be configured somewhere?

**Comment by J4sonZ** [comment]
After I deployed using nodeport
  values:
    gateways:
      istio-ingressgateway:
        injectionTemplate: gateway
        type: NodePort # add this

Istio's remote cluster can work normally, and sidecar can also be injected, but the virtual service of the host group cannot take effect. What I am more confused about is that if we use the nodeport method to deploy, which IP address is the request sent by istiod on the control side? And what is the port?

@fizos @stevenctl

**Comment by keithmattix** [comment]
@jhonatanmorais are the remote endpoints in a different network than the client cluster?

**Comment by jhonatanmorais** [comment]
Yes, @keithmattix.

I've done that only as a test alternative to the limitation found using the default service type `LoadBalancer`, explained in this [issue](https://github.com/istio/istio/issues/56355).

**Comment by lhtkk** [comment]
> 是的，我已经解决了这个问题，你可以在集群外安装像nginx这样的负载均衡器，然后把15443端口转发到
这个直接把cluster1 的nodeport ip代理到cluster2的吗
